### PR TITLE
Add --max-warnings cli flag (fixes #2939)

### LINKED ISSUE #2939: Add --max-warnings <number> flag
> Describe the issue. Is it a bug or a feature request (new rule, new option, etc.)?

Feature request. I need to run stylelint on ci. And I want ci to report that checks fail if there are any stylelint warnings. To do so, I should be able to force stylelint to exit with non-zero exit code on warnings. I think that some cli flags would be enough to do the job. For example, I use [--max-warnings 0](https://github.com/eslint/eslint/issues/2769) to do same thing in eslint.

> Which rule, if any, is this issue related to?

none

> What CSS is needed to reproduce this issue?

```css
.foo {
  color: #color;
}
```

> What stylelint configuration is needed to reproduce this issue?

```json
{
  "rules": {
    "color-no-invalid-hex": [true, {"severity": "warning"}]
  }
}
```

> Which version of stylelint are you using?

`8.0.0`

> How are you running stylelint: CLI, PostCSS plugin, Node API?

CLI with `stylelint foo.css`

> Does your issue relate to non-standard syntax (e.g. SCSS, nesting, etc.)?

no

> What did you expect to happen?

stylint exits with non-zero exit code

> What actually happened (e.g. what warnings or errors you are getting)?

stylelint exits with zero code

It seems like it is a standard use case and it should be already reported before. But I can't find anything about it. So, sorry if this is a duplicate

### PR DESCRIPTION
**Pull request by alexilyaev**
Resolves: https://github.com/stylelint/stylelint/issues/2939

A CLI flag useful in CI to exit with non zero code on warnings count greater than a specific number.  
Usually `defaultSeverity: 'warning'` is set and `--max-warnings 0` in CLI.  
That way devs don't see errors in development, but CI fails build if there are any warnings.

screen shot 2017-10-05 at 24 45 50

### CODE REVIEW

**Comment by alexilyaev** [comment]
@jeddy3 Can you guide me how to write a test for it?  
I worked with Jest in the past, just need a direction.

**Comment by alexilyaev** [comment]
Running `npm run lint` gives me this:
```
...

/Users/alex/www/Node/OSS/stylelint/system-tests/systemTestUtils.js
   1:13  error  Extra semicolon         semi
   3:28  error  Extra semicolon         semi
   4:29  error  Extra semicolon         semi
   5:40  error  Extra semicolon         semi
   8:52  error  Extra semicolon         semi
  12:50  error  Extra semicolon         semi
  16:22  error  Extra semicolon         semi
  17:51  error  Extra semicolon         semi
  23:59  error  Extra semicolon         semi
  28:54  error  Extra semicolon         semi
  30:25  error  Extra semicolon         semi
  31:5   error  Extra semicolon         semi
  35:26  error  Extra semicolon         semi
  43:14  error  Missing trailing comma  comma-dangle
  44:2   error  Extra semicolon         semi

✖ 27123 problems (27123 errors, 0 warnings)
```

I installed deps with Yarn, if it makes a difference.

**Comment by alexilyaev** [comment]
@jeddy3 Thanks.

Btw, do I need to do anything in `standalone.js`?  
What is it used for?

**Comment by jeddy3** [comment]
> Btw, do I need to do anything in standalone.js? What is it used for?

It's the main guts of the [Node API](https://github.com/stylelint/stylelint/blob/master/docs/user-guide/node-api.md#options) and where the [returned object](https://github.com/stylelint/stylelint/blob/master/docs/user-guide/node-api.md#the-returned-promise) is [constructed](https://github.com/stylelint/stylelint/blob/master/lib/standalone.js#L233-L250). A `maxWarningsExceeded` property could be added to this returned object, which `cli.js` can use.

**Comment by alexilyaev** [comment]
@jeddy3 So, I've moved the logic to `standalone.js`, adding the `maxWarningsExceeded` property, updated the docs and added a test file.

Although I'm not sure what would be the use case for using it through the Node API.

This PR is ready for a review.

File: lib/cli.js [hunk 1]
```diff
@@ -473,11 +496,22 @@ Promise.resolve()
...
+
     process.stdout.write(linted.output);
+
     if (linted.errored) {
       process.exitCode = 2;
+    } else if (maxWarnings !== undefined && linted.maxWarningsExceeded) {
+      const foundWarnings = linted.maxWarningsExceeded.foundWarnings;
+
+      process.stdout.write(
+        `Stylelint found too many warnings:${EOL}` +
```
Comment (jeddy3) [inline lib/cli.js]: Can we tweak the message to make it consistent [with how other errors (e.g. "Invalid Option" ones)](https://cloud.githubusercontent.com/assets/808227/14970067/063e5802-10be-11e6-8774-6557d80c443b.png) are formatted, please? e.g. :

```js
  process.stdout.write(
        chalk.red(`Max warnings exceeded: `) +
          `${foundWarnings} found. ` +
          chalk.dim(`${maxWarnings} allowed${EOL}${EOL}`)
      );
```

Comment (alexilyaev) [reply lib/cli.js]: @jeddy3 Looks good, updated.

**Comment by jeddy3** [comment]
@CAYdenberg If you have a moment, would you be able to advise of those flow warnings?

**Comment by CAYdenberg** [comment]
Done. @alexilyaev hope you don't mind me appending to your PR. Thanks for your contribution.

**Comment by jeddy3** [comment]
@CAYdenberg Thanks!

**Comment by alexilyaev** [comment]
@CAYdenberg Not at all, thanks.

So, merging any time soon?

**Comment by jeddy3** [comment]
@stylelint/core Can anyone spare a moment to cast a 2nd pair of eyes over this?

**Comment by jeddy3** [comment]
Change log:

> -   Added: `--max-warnings` CLI flag ([#2942](https://github.com/stylelint/stylelint/pull/2942)).

**Pull request merged**